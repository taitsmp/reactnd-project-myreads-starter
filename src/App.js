import React from 'react'
import * as BooksAPI from './BooksAPI'
import MainPage from './components/MainPage'
import SearchPage from './components/SearchPage'
import { Route } from 'react-router-dom'
import './App.css'

/*
What's left?

* rewatch the lecture on managing state and controlled components.  Did you do search right?
* rewatch any other lectures? 
* consider refactoring so that we just keep the books list in one place 
* review requirements
* you have a function to test if a network error occurred and a catch handler. Can you use these to dectect an error and render an error message to the screen?
* I occassionally get a 403 forbidden from the backend.  How best to structure code to handle this? - https://www.tjvantoll.com/2015/09/13/fetch-and-errors/
* Do you need to start this project over from a create-react-app?
* read "thinking in react" and see if you "get it".
* remove the constructor on this page.
* rewrite this in react native.  
* rewrite up the architecture. 
*/

class BooksApp extends React.Component {
  
  constructor(props) {
    super(props)

    this.state = {
      books: []
    } 

  } 

  getBook(id) {
    return this.state.books.find(b => b.id === id )
  }

  handleUpdateBook = (id, shelf) => {
    const book = {id: id, shelf: shelf}  //hack
    BooksAPI.update(book, shelf).then( shelves => {
      //shelves is a object with three lists of book ids (currently reading, want to read, read)
      console.log(shelves)
      console.log(this.getBook(id))
      
      //just do reduce?  will return a new array.
      let newBooks = this.state.books.reduce((books, b) => {
        var book

        console.log(books)
        if (b.id === id) {
          book = this.getBook(id)
          book.shelf = shelf
        }
        else
          book = b
        
        return [...books, book]
      }, []).catch(err => {
          console.log(err)
      })
      

      console.log('here' + newBooks.length)
      this.setState({books: newBooks})
    })



  }
  
  componentDidMount() {
    BooksAPI.getAll().then( obj => {
      this.setState({ books: obj })
      console.log(obj)
    }).catch(err => {
        if (err instanceof TypeError)
          console.log("Network Not Present")
        else
          throw err
      })
  }  

  render() {
    //had to print the state here not in componentDidMount()
    console.log(this.state);
    return (
      <div className="app">
        <Route path="/search" render={() => (
          <SearchPage onUpdateBook={this.handleUpdateBook} />
        )} />
        <Route exact path='/' render={() => (
          <MainPage books={this.state.books} onUpdateBook={this.handleUpdateBook} />
        )} />
       
      </div>
    )
  }
}

export default BooksApp
