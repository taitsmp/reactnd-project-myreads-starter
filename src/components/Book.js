import React, { Component } from 'react'
import PropTypes from 'prop-types'

//this is going to just have a render method and props.
class Book extends Component {

    render() {
        const coverStyle = {
            width: 128,
            height: 193,
            backgroundImage: "url('"+this.props.thumbnail+"')"
        }
        return (
                        <div className="book">
                          <div className="book-top">
                            <div className="book-cover" style={coverStyle}></div>
                            <div className="book-shelf-changer">
                              <select>
                                <option value="none" disabled>Move to...</option>
                                <option value="currentlyReading">Currently Reading</option>
                                <option value="wantToRead">Want to Read</option>
                                <option value="read">Read</option>
                                <option value="none">None</option>
                              </select>
                            </div>
                          </div>
                          <div className="book-title">{this.props.title}</div>
                          <div className="book-authors">{this.props.author}</div>
                        </div>
        )
    }
        

}

//leaving this outside class definition in case I change this to a propType
Book.PropTypes = {
    title: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    thumbnail: PropTypes.string.isRequired
}

export default Book