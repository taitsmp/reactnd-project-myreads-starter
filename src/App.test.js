import React from 'react'
import ReactDOM from 'react-dom'
import { MemoryRouter as Router} from 'react-router-dom' // 4.0.0
import App from './App'
import Book from './components/Book'
import renderer from 'react-test-renderer';


it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<Router><App /></Router>, div)
})

// - https://facebook.github.io/jest/docs/snapshot-testing.html
// - Functional components should be testable here.  
// - bad idea? will change too much?

it('renders a book correctly', () => {
  const book = {  title: "The Linux Command Line",
                thumbnail: "http://books.google.com/books/content?id=nggnmAEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api",
                id: "nggnmAEACAAJ",
                shelf: "currentlyReading",
      }
  const onUpdateBook = () => {}
  const tree = renderer.create(
    <Book 
    thumbnail={book.thumbnail} 
    title={book.title} 
    author={book.author} 
    shelf={book.shelf}
    id={book.id}
    onUpdateBook={onUpdateBook} />
  ).toJSON();
  expect(tree).toMatchSnapshot();
})
