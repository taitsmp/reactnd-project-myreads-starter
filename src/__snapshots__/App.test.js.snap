exports[`test renders a book correctly 1`] = `
<div
  className="book">
  <div
    className="book-top">
    <div
      className="book-cover"
      style={
        Object {
          "backgroundImage": "url(\'http://books.google.com/books/content?id=nggnmAEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api\')",
          "height": 193,
          "width": 128,
        }
      } />
    <div
      className="book-shelf-changer">
      <select
        onChange={[Function]}
        value="currentlyReading">
        <option
          disabled={true}
          value="none">
          Move to...
        </option>
        <option
          value="currentlyReading">
          Currently Reading
        </option>
        <option
          value="wantToRead">
          Want to Read
        </option>
        <option
          value="read">
          Read
        </option>
        <option
          value="none">
          None
        </option>
      </select>
    </div>
  </div>
  <div
    className="book-title">
    The Linux Command Line
  </div>
  <div
    className="book-authors" />
</div>
`;
